### ISSUE TITLE
A declarative secure-by-default headless weechat

### ISSUE THREAD

**Issue by tomfitzhenry** [NixOS/nixpkgs#85080]
_tl;dr `services.weechat` isn't declarative and is hard-to-make secure. I'm working on it. This bug is how I'm tracking it._

`services.weechat` has a few problems:
* the config in `/var/lib/weechat/` is mutated by weechat and the user (who issues commands such as `set foo bar`, and is thus impure.
* the only (?) supported way to [configure weechat in NixOS is via `--run-command`](https://github.com/NixOS/nixpkgs/blob/aad81ec8cd158a9e11f4d35fc6022d1300fcb4c6/doc/builders/packages/weechat.xml#L43-L47), which executes on startup, after the mutable config is read from disk, and thus a robust config has to undo any potential unwanted config that has been written to disk
* logs are written to `/var/lib/weechat` rather than stdout and `/var/log/weechat`
* it's difficult to securely enable the relay
    * Weechat enables plugins that perform exec by default, such as `exec`, `ruby`, `python`, `guile`, etc.
    * the relay happily relays exec commands
    * Thus, the relay will perform remote code execution protected by the relay's authentication. I have no evidence this is exploitable, but it doesn't have as many eyes as typical authentication implementations (OpenSSH, OpenSSL), and doesn't exhibit defense-in-depth.
    * The risk can be mitigated with a plugin allow list (`weechat.plugin.autoload`) and a relay command allow list (`relay.weechat.commands`), but even if you have suitable values for these, the aforementioned impure config makes it difficult to have confidence this config remains working.
* it's difficult to sandbox, since a typical weechat installation as a mutable-anything-goes box.

I propose `services.weechat-headless`:
* declarative Weechat config in NixOS: we'll generate the config files (`irc.conf`, `relay.conf`, etc.)
* immutable config (Weechat will be prohibited from writing over it)
* secure-by-default
    * only core plugins enabled by default
    * a documented way to enable the relay, and the relay command allow list.
    * a `DynamicUser=` sandbox, and more hardening.
    * no local access via `screen`

I have a working prototype of this, and came across a few issues:
* Weechat documentation advises users to use `/set` rather than write config files by hand.
* Weechat's config parser is sensitive (e.g. `foo = "bar"` will parse, `foo="bar"` will not)
* Weechat's config parser will ignore parse errors and fallback to default values.
* Weechat writes its main log (`weechat.log`) to the same directory as its config files. This blocks us being able to make the config read-only, which is how we preserve purity.

Here's the plan:

- [x] Patch Weechat upstream to log to stdout, if running in headless mode. https://github.com/weechat/weechat/issues/1475
- [ ] Write `services.weechat-headless` (is `weechat-relay` a better name? It's more descriptive of its intended usecase.)
- [ ] Document how to enable the relay on `services.weechat-headless`.
- [ ] Optional: Patch Weechat upstream to treat config parse errors as failures, if running in headless mode.
- [ ] Optional: Write a function to convert a Nix attrset into the ini-_like_ format that Weechat expects. This will make the module easier to use, and mean users are less likely to have invalid config.
- [x] Optional: Patch Weechat upstream to reload config files (i.e. `/reload`) on SIGHUP, so when a user changes their Weechat config in NixOS, their `weechat-headless.service` can be reloaded rather than restarted. https://github.com/weechat/weechat/issues/1476
- [ ] Optional: Figure out how to handle buffer unread state tracking, which is typically stored in config files. When those config files are read-only, unread state will not persist across restarts.

**Comment by tilpner** [comment]
I've used a declaratively-configured weechat in the past, it wasn't that fun. It's designed to have settings changed interactively, and forcing a rebuild every time you join a new channel is annoying (for per-buffer settings). Perhaps it gets more ergonomic if you keep it partially mutable (at least for buffer notification settings).

The following two files are the config I no longer use, scavenge whatever you consider helpful (they are [MIT-licensed](https://github.com/NixOS/nixpkgs/blob/master/COPYING)).
<details><summary><b>default.nix</b></summary><p>


```nix
{ lib, callPackage, runCommand, writeTextDir, symlinkJoin, makeBubblewrap,
  fetchFromGitHub, weechat, pass }:

with lib;
let
  flattenAttrs = set:
    let recurse = k: v:
      if isAttrs v
      then mapAttrsToList (k': recurse "${k}.${k'}") v
      else nameValuePair k v;
    in listToAttrs (flatten (mapAttrsToList recurse set));
  
  mkValueString = v:
    if isInt v then toString v
    else if isString v then "\"${v}\""
    else if true == v then "on"
    else if false == v then "off"
    else abort "unsupported value type: ${builtins.typeOf v}";

  mkKeyValue = k: v:
    if isAttrs v then concatStringsSep "\n" (mapAttrsToList (k': mkKeyValue "${k}.${k'}") (flattenAttrs v))
    else "${k} = ${mkValueString v}";

  toConfigFile = name: attrs:
    writeTextDir "${name}.conf" (generators.toINI { inherit mkKeyValue; } attrs);

  scripts = fetchFromGitHub {
    owner = "weechat";
    repo = "scripts";
    rev = "712b6773d1ae49771b5030341433fb55dd89a6d6";
    sha256 = "17bkgdvavwz4y7wica9vfhaygibwwppn6f09cbhh1vad68a1s4rs";
  };

  scriptsDir = runCommand "scripts" {} ''
    mkdir -p $out/python/autoload
    ln -s \
      ${scripts}/python/go.py \
      ${scripts}/python/autosort.py \
      ${scripts}/python/prism.py \
      $out/python/autoload/

    mkdir -p $out/perl/autoload
    ln -s \
      ${scripts}/perl/colorize_lines.pl \
      ${scripts}/perl/notify_send.pl \
      $out/perl/autoload/
  '';

  config = import ./config.nix { inherit lib; };
  homeDir = symlinkJoin {
    name = "weechat-home";
    paths = [ scriptsDir ] ++ (mapAttrsToList toConfigFile config);
  };
in makeBubblewrap {
  name = "irc";
  argv = "${weechat}/bin/weechat";

  env.HOME = "/jail";
  env.ZNC_PASSWORD = "$(${pass}/bin/pass some/password)";

  rules = p: with p; [
    base
    (tmpfs "/jail")
  ];

  preRunInside = ''
    cp -r ${homeDir} /jail/.weechat
    chmod -R 700 /jail
  '';
} // { passthru = { inherit homeDir; }; }

```

</p></details>

<details><summary><b>config.nix</b></summary><p>

```nix
{ lib }:

with lib; {
  irc.server_default = {
    capabilities = concatStringsSep "," [
      "account-notify" "away-notify" "cap-notify" "multi-prefix" "server-time"
      "znc.in/server-time-iso" "znc.in/self-message"
    ];
    nicks = "Foo,Bar,Quux";
    realname = "Quuux";
    ssl_verify = false;
    username = "bar";
  };

  irc.server = let
    zncNetwork = name: {
      addresses = "znc.bar.com";
      username = "bar/${name}";
      password = "\${env:ZNC_PASSWORD}";
      ssl = true;
      ssl_verify = false;
      autoconnect = false;
    }; 
    networks = [ "freenode" "mozilla" "oftc" ];
  in (listToAttrs (map (n: nameValuePair n (zncNetwork n)) networks));

  irc.look = {
    buffer_switch_autojoin = false;
    buffer_switch_join = false;
    server_buffer = "independent";
    smart_filter_delay = 60;
    temporary_servers = true;
    lag_reconnect = 0;
  };

  weechat = {
    notify.irc =
      let s = {
        freenode = [ "#nixos" ];
      };
      in mapAttrs (k: v: listToAttrs (map (v: nameValuePair v "message") v)) s;

    look = {
      prefix_align_max = 15;
      color_nick_offline = true;
      buffer_time_format = "\${color:252}%H\${color:245}%M\${color:240}%S";
      buffer_notify_default = "highlight";
    };

    plugin.autoload = "*,!aspell,!fifo,!relay,!xfer,!tcl,!guile";

    completion.nick_completer = " -";

    filter.irc_smart = "on;*;irc_smart_filter;*";
    filter.nixos_github = "on;*;nick_nixos_github+irc_join;*";

    color = {
      bar_more = "lightmagenta";
      chat = "default";
      chat_bg = "default";

      chat_nick_colors = concatStringsSep "," [
        "cyan" "magenta" "green" "brown" "lightcyan" "lightmagenta" "blue"
        "31" "35" "38" "40" "49" "70" "80" "99" "112" "138" "142" "148" "167"
        "169" "174" "176" "178" "184" "186" "210" "212" "215" "247"
      ];

      chat_highlight = "red";
      chat_highlight_bg = "black";

      chat_nick_offline = "!default";
      chat_nick_offline_highlight = "!default";
      chat_nick_offline_highlight_bg = "default";
    };

    bar = {
      buflist = {
        color_bg = "default";
        color_delim = "default";
        color_fg = "default";
        conditions = "";
        filling_left_right = "vertical";
        filling_top_bottom = "columns_vertical";
        hidden = false;
        items = "buflist";
        position = "right";
        priority = 0;
        separator = true;
        size = 0;
        size_max = 0;
        type = "root";
      };

      fset = {
        color_bg = "default";
        color_delim = "cyan";
        color_fg = "default";
        conditions = "\${buffer.full_name} == fset.fset";
        filling_left_right = "vertical";
        filling_top_bottom = "horizontal";
        hidden = false;
        items = "fset";
        position = "top";
        priority = 0;
        separator = true;
        size = 3;
        size_max = 3;
        type = "window";
      };

      input = {
        color_bg = "default";
        color_delim = "cyan";
        color_fg = "default";
        conditions = "";
        filling_left_right = "vertical";
        filling_top_bottom = "horizontal";
        hidden = false;
        items = "[input_prompt]+(away),[input_search],[input_paste],input_text";
        position = "bottom";
        priority = 1000;
        separator = false;
        size = 0;
        size_max = 0;
        type = "window";
      };

      isetbar = {
        color_bg = "default";
        color_delim = "cyan";
        color_fg = "default";
        conditions = "";
        filling_left_right = "vertical";
        filling_top_bottom = "horizontal";
        hidden = true;
        items = "isetbar_help";
        position = "top";
        priority = 0;
        separator = true;
        size = 3;
        size_max = 3;
        type = "window";
      };

      nicklist = {
        color_bg = "default";
        color_delim = "cyan";
        color_fg = "default";
        conditions = "\${nicklist}";
        filling_left_right = "vertical";
        filling_top_bottom = "columns_vertical";
        hidden = true;
        items = "buffer_nicklist";
        position = "right";
        priority = 200;
        separator = true;
        size = 0;
        size_max = 0;
        type = "window";
      };

      status = {
        color_bg = 0;
        color_delim = "cyan";
        color_fg = "default";
        conditions = "";
        filling_left_right = "vertical";
        filling_top_bottom = "horizontal";
        hidden = false;
        items = "[time],[buffer_last_number],[buffer_plugin],buffer_number+:+buffer_name+(buffer_modes)+{buffer_nicklist_count}+buffer_zoom+buffer_filter,scroll,[lag],[hotlist],completion";
        position = "bottom";
        priority = 500;
        separator = false;
        size = 1;
        size_max = 0;
        type = "window";
      };

      title = {
        color_bg = 0;
        color_delim = "cyan";
        color_fg = "default";
        conditions = "";
        filling_left_right = "vertical";
        filling_top_bottom = "horizontal";
        hidden = false;
        items = "buffer_title";
        position = "top";
        priority = 500;
        separator = false;
        size = 1;
        size_max = 0;
        type = "window";
      };
    };

    key = {
      "meta-#" = "/go";
      "meta-e" = "/go freenode.#nixos";

      "ctrl-?" = "/input delete_previous_char";
      "ctrl-A" = "/input insert \\x01";
      "ctrl-B" = "/input insert \\x02";
      "ctrl-C" = "/input insert \\x03";
      "ctrl-C_" = "/input insert \\x1F";
      "ctrl-Cb" = "/input insert \\x02";
      "ctrl-Cc" = "/input insert \\x03";
      "ctrl-Ci" = "/input insert \\x1D";
      "ctrl-Co" = "/input insert \\x0F";
      "ctrl-Cv" = "/input insert \\x16";
      "ctrl-D" = "/input delete_next_char";
      "ctrl-E" = "/input move_end_of_line";
      "ctrl-F" = "/input move_next_char";
      "ctrl-G" = "/input insert \\x0303\\x03\\x03; /input move_previous_char; /input move_previous_char";
      "ctrl-H" = "/input delete_previous_char";
      "ctrl-I" = "/input complete_next";
      "ctrl-J" = "/input return";
      "ctrl-K" = "/input delete_end_of_line";
      "ctrl-L" = "/window refresh";
      "ctrl-M" = "/input return";
      "ctrl-N" = "/buffer +1";
      "ctrl-P" = "/buffer -1";
      "ctrl-R" = "/input search_text_here";
      "ctrl-S" = "/input insert \\x0303$ \\x03\\x03; /input move_previous_char; /input move_previous_char";
      "ctrl-Sctrl-U" = "/input set_unread";
      "ctrl-T" = "/input transpose_chars";
      "ctrl-U" = "/input delete_beginning_of_line";
      "ctrl-W" = "/input delete_previous_word";
      "ctrl-X" = "/input switch_active_buffer";
      "ctrl-Y" = "/input clipboard_paste";
      "meta-meta-OP" = "/bar scroll buflist * b";
      "meta-meta-OQ" = "/bar scroll buflist * e";
      "meta-meta2-1~" = "/window scroll_top";
      "meta-meta2-23~" = "/bar scroll nicklist * b";
      "meta-meta2-24~" = "/bar scroll nicklist * e";
      "meta-meta2-4~" = "/window scroll_bottom";
      "meta-meta2-5~" = "/window scroll_up";
      "meta-meta2-6~" = "/window scroll_down";
      "meta-meta2-7~" = "/window scroll_top";
      "meta-meta2-8~" = "/window scroll_bottom";
      "meta-meta2-A" = "/buffer -1";
      "meta-meta2-B" = "/buffer +1";
      "meta-meta2-C" = "/buffer +1";
      "meta-meta2-D" = "/buffer -1";
      "meta--" = "/filter toggle @";
      "meta-/" = "/input jump_last_buffer_displayed";
      "meta-0" = "/buffer *10";
      "meta-1" = "/buffer *1";
      "meta-2" = "/buffer *2";
      "meta-3" = "/buffer *3";
      "meta-4" = "/buffer *4";
      "meta-5" = "/buffer *5";
      "meta-6" = "/buffer *6";
      "meta-7" = "/buffer *7";
      "meta-8" = "/buffer *8";
      "meta-9" = "/buffer *9";
      "meta-<" = "/input jump_previously_visited_buffer";
      "meta-=" = "/filter toggle";
      "meta->" = "/input jump_next_visited_buffer";
      "meta-M" = "/window scroll_up";
      "meta-N" = "/windows scroll_down";
      "meta-OA" = "/input history_global_previous";
      "meta-OB" = "/input history_global_next";
      "meta-OC" = "/input move_next_word";
      "meta-OD" = "/input move_previous_word";
      "meta-OF" = "/input move_end_of_line";
      "meta-OH" = "/input move_beginning_of_line";
      "meta-OP" = "/bar scroll buflist * -100%";
      "meta-OQ" = "/bar scroll buflist * +100%";
      "meta-Oa" = "/input history_global_previous";
      "meta-Ob" = "/input history_global_next";
      "meta-Oc" = "/input move_next_word";
      "meta-Od" = "/input move_previous_word";
      "meta2-15~" = "/buffer -1";
      "meta2-17~" = "/buffer +1";
      "meta2-18~" = "/window -1";
      "meta2-19~" = "/window +1";
      "meta2-1;3A" = "/buffer -1";
      "meta2-1;3B" = "/buffer +1";
      "meta2-1;3C" = "/buffer +1";
      "meta2-1;3D" = "/buffer -1";
      "meta2-1;3F" = "/window scroll_bottom";
      "meta2-1;3H" = "/window scroll_top";
      "meta2-1;5A" = "/input history_global_previous";
      "meta2-1;5B" = "/input history_global_next";
      "meta2-1;5C" = "/input move_next_word";
      "meta2-1;5D" = "/input move_previous_word";
      "meta2-1~" = "/input move_beginning_of_line";
      "meta2-200~" = "/input paste_start";
      "meta2-201~" = "/input paste_stop";
      "meta2-20~" = "/bar scroll title * -30%";
      "meta2-21~" = "/bar scroll title * +30%";
      "meta2-23;3~" = "/bar scroll nicklist * b";
      "meta2-23~" = "/bar scroll nicklist * -100%";
      "meta2-24;3~" = "/bar scroll nicklist * e";
      "meta2-24~" = "/bar scroll nicklist * +100%";
      "meta2-3~" = "/input delete_next_char";
      "meta2-4~" = "/input move_end_of_line";
      "meta2-5;3~" = "/window scroll_up";
      "meta2-5~" = "/window page_up";
      "meta2-6;3~" = "/window scroll_down";
      "meta2-6~" = "/window page_down";
      "meta2-7~" = "/input move_beginning_of_line";
      "meta2-8~" = "/input move_end_of_line";
      "meta2-A" = "/input history_previous";
      "meta2-B" = "/input history_next";
      "meta2-C" = "/input move_next_char";
      "meta2-D" = "/input move_previous_char";
      "meta2-F" = "/input move_end_of_line";
      "meta2-G" = "/window page_down";
      "meta2-H" = "/input move_beginning_of_line";
      "meta2-I" = "/window page_up";
      "meta2-Z" = "/input complete_previous";
      "meta2-[E" = "/buffer -1";
      "meta-_" = "/input redo";
      "meta-b" = "/input move_previous_word";
      "meta-d" = "/input delete_next_word";
      "meta-f" = "/input move_next_word";
      "meta-h" = "/input hotlist_clear";
      "meta-jmeta-f" = "/buffer -";
      "meta-jmeta-l" = "/buffer +";
      "meta-jmeta-r" = "/server raw";
      "meta-jmeta-s" = "/server jump";
      "meta-k" = "/input grab_key_command";
      "meta-l" = "/window bare";
      "meta-m" = "/window scroll_up";
      "meta-n" = "/window scroll_down";
      "meta-p" = "/window scroll_previous_highlight";
      "meta-return" = "/input return";
      "meta-u" = "/window scroll_unread";
      "meta-wmeta-meta2-A" = "/window up";
      "meta-wmeta-meta2-B" = "/window down";
      "meta-wmeta-meta2-C" = "/window right";
      "meta-wmeta-meta2-D" = "/window left";
      "meta-wmeta2-1;3A" = "/window up";
      "meta-wmeta2-1;3B" = "/window down";
      "meta-wmeta2-1;3C" = "/window right";
      "meta-wmeta2-1;3D" = "/window left";
      "meta-wmeta-b" = "/window balance";
      "meta-wmeta-s" = "/window swap";
      "meta-x" = "/input zoom_merged_buffer";
      "meta-z" = "/window zoom";
      "ctrl-_" = "/input undo";
    };

    key_search = {
      "ctrl-I" = "/input search_switch_where";
      "ctrl-J" = "/input search_stop_here";
      "ctrl-M" = "/input search_stop_here";
      "ctrl-Q" = "/input search_stop";
      "ctrl-R" = "/input search_switch_regex";
      "meta2-A" = "/input search_previous";
      "meta2-B" = "/input search_next";
      "meta-c" = "/input search_switch_case";
    };

    key_cursor = {
      "ctrl-J" = "/cursor stop";
      "ctrl-M" = "/cursor stop";
      "meta-meta2-A" = "/cursor move area_up";
      "meta-meta2-B" = "/cursor move area_down";
      "meta-meta2-C" = "/cursor move area_right";
      "meta-meta2-D" = "/cursor move area_left";
      "meta2-1;3A" = "/cursor move area_up";
      "meta2-1;3B" = "/cursor move area_down";
      "meta2-1;3C" = "/cursor move area_right";
      "meta2-1;3D" = "/cursor move area_left";
      "meta2-A" = "/cursor move up";
      "meta2-B" = "/cursor move down";
      "meta2-C" = "/cursor move right";
      "meta2-D" = "/cursor move left";
      "@chat:Q" = "hsignal:chat_quote_time_prefix_message;/cursor stop";
      "@chat:m" = "hsignal:chat_quote_message;/cursor stop";
      "@chat:q" = "hsignal:chat_quote_prefix_message;/cursor stop";
    };
  };

  alias.cmd = {
    alis = "/msg alis list $*";
    cs = "/msg ChanServ $*";
    ns = "/msg NickServ $*";
    znc = "/msg *status $*";

    spoiler = "/msg $channel 1,1$*";
    bhide = "/set weechat.bar.buffers.hidden toggle";
    voice = "/msg ChanServ voice $channel $*";
    devoice = "/msg ChanServ devoice $channel $*";
  };

  trigger.trigger = {
    input_command_color = {
      arguments = "input_text_display";
      command = "";
      conditions = "\${tg_string} =~ ^/($|[^/])";
      enabled = true;
      hook = "modifier";
      post_action = "none";
      regex = "#/(.+)#\${color:39}/\${color:74}\${re:1}#";
      return_code = "ok";
    };

    last_nick = {
      arguments = "irc.*;notify_message";
      command = "/buffer set localvar_set_last_nick \${tg_tag_nick}";
      conditions = "\${tg_displayed} && \${type} == channel && \${tg_tag_nick} != \${nick}";
      enabled = true;
      hook = "print";
      post_action = "none";
      regex = "";
      return_code = "ok";
    };

    complete_last_nick = {
      arguments = "/input complete_next";
      command = "/input insert \${last_nick}";
      conditions = "\${type} == channel && \${buffer.input_buffer_length} == 0 && \${last_nick}";
      enabled = true;
      hook = "command_run";
      post_action = "none";
      regex = "";
      return_code = "ok";
    };

    space_cmd = {
      arguments = "input_text_for_buffer";
      command = "/print \${tg_string_nocolor} prevented";
      conditions = "\${tg_string} =~ ^\\s+/";
      enabled = true;
      hook = "modifier";
      post_action = "none";
      regex = "/.*//";
      return_code = "ok";
    };
  };

  autosort = {
    sorting.replacements = ''[["##", "#"]]'';
    sorting.rules = ''[["core", 0], ["irc", 10], ["*", 1], ["irc.irc_raw", 0], ["irc.server", 1]]'';
  };

  buflist.format = {
    buffer = "\${format_number}\${indent}\${format_nick_prefix}\${color_hotlist}\${name}";
    buffer_current = "\${color:,red}\${format_buffer}";
    hotlist_highlight = "\${color:magenta,black}";
    hotlist_message = "\${color:blue,black}";
    hotlist_private = "\${color:green,black}";
    indent = " ";
  };

  plugins.var.perl = {
    colorize_lines = {
      buffers = "all";
      highlight = true;
      ignore_tags = "irc_ctcp";
      lines = "on";
      own_lines = "on";
      tags = "irc_privmsg";
    };

    notify_send = {
      command = "notify-send $name \"$message\"";
      wait_highlight = 60;
      wait_pm = 180;
    };
  };

  plugins.var.python.go = {
    auto_jump = "off";
    color_name = "black,cyan";
    color_name_highlight = "red,cyan";
    color_name_highlight_selected = "red,brown";
    color_name_selected = "black,brown";
    color_number = "yellow,magenta";
    color_number_selected = "yellow,red";
    message = "Go to: ";
    short_name = "off";
    sort = "number,beginning";
    use_core_instead_weechat = "off";
  };
}
```

</p></details>

**Comment by tomfitzhenry** [comment]
> I've used a declaratively-configured weechat in the past, it wasn't that fun.

@tilpner , I'm glad you've tried this and have shared your experience and code. Thanks!

> It's designed to have settings changed interactively, and forcing a rebuild every time you join a new channel is annoying (for per-buffer settings).

Once https://github.com/weechat/weechat/issues/1476 is merged, weechat will reload on SIGHUP, and thus config changes can easily be enacted in Weechat via the usual NixOS/systemd workflow. This will work well for many settings, but won't be perfect. For example, I expect changing autojoin settings and issuing /reload doesn't trigger a join.

Ad-hoc joins (via `/join`) will continue to work.

> Perhaps it gets more ergonomic if you keep it partially mutable (at least for buffer notification settings).

Can you explain a bit more why buffer notification settings would particularly benefit from being mutable?

One of the difficulties with part-mutability is the config is read-only. An immutable config comes with benefits and costs. `services.weechat-headless` will be more suited to those who prefer immutable config.

> The following two files are the config I no longer use, scavenge whatever you consider helpful (they are MIT-licensed).

Those look great! They're what I meant by "Write a function to convert a Nix attrset into the ini-like format that Weechat expects.". Thanks for that.

**Comment by tomfitzhenry** [comment]
I don't think this should immediately go into NixOS, because of the uncertain benefits of immutable config, in light of @tilpner's experience. I will provide this module elsewhere, and if people are interested, we can add it to NixOS later.

**Comment by tilpner** [comment]
> [...] weechat will reload on SIGHUP, and thus config changes can easily be enacted in Weechat via the usual NixOS/systemd workflow. [...]

Neat, this should help a lot.

> Can you explain a bit more why buffer notification settings would particularly benefit from being mutable?

There's nothing special about them, it was just an example. I frequently change per-buffer notification levels with aliases (`/subscribe`, `/unsubscribe`), and would like those changes to persist past restarts.

I'm not necessarily against adding this, but I encourage you (and other potential users) to test-drive it for a few months. As I implemented it, it was a pain to use, but the SIGHUP reloading without restart might make it Good Enough™.

**Comment by tomfitzhenry** [comment]
NixOS 20.09 includes weechat 2.9, which supports the `--stdout` flag, so weechat-headless can output via journald. This allows us to make the weechat directory read-only/transient (either way stops weechat-headless from overwriting config and thus preserves purity), while preserving stdout-style logging.

> I'm not necessarily against adding this, but I encourage you (and other potential users) to test-drive it for a few months. As I implemented it, it was a pain to use, but the SIGHUP reloading without restart might make it Good Enough™.

I agree, and so I'll close this. If I ever get around to publishing my proposed weechat-headless, I'll link it here.

For now, here's what it looks like in nixos 20.09: https://gist.github.com/tomfitzhenry/40f929dbd0864b09f9bc3b5a69cba1e3
